core/arch/atomic_exports.h

```diff
@@ -936,11 +936,12 @@ atomic_compare_exchange_int(volatile int *var, int compare, int exchange)
      * sequentially consistent memory order semantics.
      */
     __asm__ __volatile__("fence   iorw,ow\n\t"
-                         "1: lr.w.aq t0, (%1)\n\t"
-                         "   bne     t0, %2, 2f\n\t"
-                         "   sc.w.aq %0, %3, (%1)\n\t"
-                         "   bnez    %0, 1b\n\t"
-                         "2:"
+                         "1: lr.w.aqrl  t0, (%1)\n\t"
+                         "   bne        t0, %2, 2f\n\t"
+                         "   sc.w.rl    %0, %3, (%1)\n\t"
+                         "   bnez       %0, 1b\n\t"
+                         "2:\n\t"
+                         "   subw       %0, t0, %2\n\t"
```

Comment:
Why is this `subw` needed? Without it the logic seems to work fine:

* if `[var] != compare`, then `res` remains 1, so we return false below, as expected.
* if `[var] == compare`, then `res` will be zero when the `lr-sc` loop completes successfully, so we return true below as expected.

This `subw` seems to only redundantly set `res`.

Am I missing something?

Reply:
Indeed, the `subw` is redundant. I forgot we initialized res as 1 already, the `subw` is only needed when res is not initialized.

Reply:
Fixed in https://github.com/DynamoRIO/dynamorio/pull/6089/commits/3be7fd4d7b709206f05d900faa031773afc68cc5

Reply:
We should also fix it in atomic_compare_exchange_int64.

Reply:
Fixed in https://github.com/DynamoRIO/dynamorio/pull/6089/commits/0cdef8be9747ba3903a0b771cab6d841964cd6e5.